frame/dex/src/lib.rs

```diff
@@ -0,0 +1,855 @@
+		/// Amount can't be zero.
+		ZeroAmount,
+		/// Calculated amount out is less than min desired.
+		InsufficientOutputAmount,
+		/// Insufficient liquidity in the pool.
+		InsufficientLiquidity,
+		/// Excessive input amount.
+		ExcessiveInputAmount,
+		/// Only pools with native on one side are valid.
+		PoolMustContainNativeCurrency,
+		/// The provided path must consists of 2 assets at least.
+		InvalidPath,
+		/// It was not possible to calculate path data.
+		PathError,
+	}
+
+	// Pallet's callable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates an empty liquidity pool and an associated new `lp_token` asset
+		/// (the id of which is returned in the `Event::PoolCreated` event).
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::create_pool())]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			asset1: MultiAssetId<T::AssetId>,
+			asset2: MultiAssetId<T::AssetId>,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			ensure!(asset1 != asset2, Error::<T>::EqualAssets);
+
+			let pool_id = Self::get_pool_id(asset1, asset2);
+			let (asset1, _asset2) = pool_id;
+
+			if !T::AllowMultiAssetPools::get() && asset1 != MultiAssetId::Native {
+				Err(Error::<T>::PoolMustContainNativeCurrency)?;
+			}
+
+			ensure!(!Pools::<T>::contains_key(&pool_id), Error::<T>::PoolExists);
+
+			let pool_account = Self::get_pool_account(pool_id);
+			let lp_token = NextPoolAssetId::<T>::get().unwrap_or(T::PoolAssetId::initial_value());
+
+			let next_lp_token_id = lp_token.increment();
+			NextPoolAssetId::<T>::set(Some(next_lp_token_id));
+
+			T::PoolAssets::create(lp_token, pool_account.clone(), true, MIN_LIQUIDITY.into())?;
+
+			let pool_info = PoolInfo { lp_token };
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolCreated { creator: sender, pool_id, lp_token });
+
+			Ok(())
+		}
+
+		/// Provide liquidity into the pool of `asset1` and `asset2`.
+		/// NOTE: an optimal amount of asset1 and asset2 will be calculated and
+		/// might be different to provided `amount1_desired`/`amount2_desired`
+		/// thus it's needed to provide the min amount you're happy to provide.
+		/// Params `amount1_min`/`amount2_min` represent that.
+		/// `mint_to` will be sent the liquidity tokens that represent this share of the pool.
+		/// `keep_alive` true will fail the transaction if in enacting the transaction
+		/// would take the sender's balance below the existential deposit.
+		/// `deadline` is the blocknumber until which you are happy for the transaction to occur.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::add_liquidity())]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			asset1: MultiAssetId<T::AssetId>,
+			asset2: MultiAssetId<T::AssetId>,
+			amount1_desired: AssetBalanceOf<T>,
+			amount2_desired: AssetBalanceOf<T>,
+			amount1_min: AssetBalanceOf<T>,
+			amount2_min: AssetBalanceOf<T>,
+			mint_to: T::AccountId,
+			deadline: T::BlockNumber,
```

Comment:
there is no need to have deadline as we already have transaction mortality

Reply:
Thanks - it makes sense that this functionality is generic and available to every transaction.